### PR TITLE
Enable security information mapping for RoleSecurityHandler

### PR DESCRIPTION
**Pull request by core23**
## Subject

I am targeting this branch, because this feature is BC.

Finally provide #4925 feature

## Changelog

```markdown
### Added
-  Enable security information mapping for `RoleSecurityHandler`
```

If you want to combine multiple permissions, you can now use the `sonata.admin.security.handler.role` for this.

```yaml
sonata_admin:
    security:
        handler:    sonata.admin.security.handler.role

        information:
            VIEWER:    [VIEW, LIST, EXPORT]
            EDITOR:    [EDIT, LIST, CREATE]
            ADMIN:     [OPERATOR, MASTER]
```

If you have a user with a `ROLE_ACME_ADMIN_VIEWER` permission, the security handler will work if your admin asks for VIEW, LIST or EXPORTER grants.

The changes are BC, because if will also work if you have no special information mapping.

### CODE REVIEW

File: src/Security/Handler/RoleSecurityHandler.php [hunk 1]
```diff
@@ -125,4 +116,52 @@ private function isAnyGranted(array $attributes, ?object $subject = null): bool
...
+        // NEXT_MAJOR: Change the foreach to a single check.
+        $useAll = false;
+        foreach ($attributes as $attribute) {
+            // If the attribute is not already a ROLE_ we generate the related role.
+            if (\is_string($attribute) && !str_starts_with($attribute, 'ROLE_')) {
+                $useAll = true;
+            }
+        }
+
+        return $useAll;
```
Comment (phansys) [inline src/Security/Handler/RoleSecurityHandler.php]:
```suggestion
        // NEXT_MAJOR: Change the foreach to a single check.
        foreach ($attributes as $attribute) {
            // If the attribute is not already a ROLE_ we generate the related role.
            if (\is_string($attribute) && !str_starts_with($attribute, 'ROLE_')) {
                return true;
            }
        }

        return false;
```

File: src/Security/Handler/RoleSecurityHandler.php [hunk 2]
```diff
@@ -125,4 +116,52 @@ private function isAnyGranted(array $attributes, ?object $subject = null): bool
 
         return false;
     }
+
+    /**
+     * @param array<string|Expression> $attributes
+     */
+    private function hasAllRule(mixed $attributes): bool
```
Comment (phansys) [inline src/Security/Handler/RoleSecurityHandler.php]: Is "rule" the right term here?

Comment (core23) [reply src/Security/Handler/RoleSecurityHandler.php]: Thanks for the review. Adjusted the name

File: src/Security/Handler/RoleSecurityHandler.php [hunk 3]
```diff
@@ -125,4 +116,51 @@ private function isAnyGranted(array $attributes, ?object $subject = null): bool
 
         return false;
     }
+
+    /**
+     * @param array<string|Expression> $attributes
+     */
+    private function hasOnlyAdminRoles(mixed $attributes): bool
```
Comment (VincentLanglet) [inline src/Security/Handler/RoleSecurityHandler.php]:
```suggestion
    private function hasOnlyAdminRoles(array $attributes): bool
```
according to the phpdoc

File: src/Security/Handler/RoleSecurityHandler.php [hunk 4]
```diff
@@ -125,4 +116,51 @@ private function isAnyGranted(array $attributes, ?object $subject = null): bool
...
+        return false;
+    }
+
+    /**
+     * @param array<string|Expression> $attributes
+     * @param AdminInterface<object>   $admin
+     *
+     * @return array<string|Expression>
+     */
+    private function mapAttributes(mixed $attributes, AdminInterface $admin): array
```
Comment (VincentLanglet) [inline src/Security/Handler/RoleSecurityHandler.php]:
```suggestion
    private function mapAttributes(array $attributes, AdminInterface $admin): array
```

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

**Review by VincentLanglet — APPROVED** [review]

**Review by phansys — APPROVED** [review]

**Comment by core23** [comment]
phpstan is also failing on the dev-kit PRs :(

https://github.com/sonata-project/SonataAdminBundle/actions/runs/9930955039/job/27430272674?pr=8195

**Pull request merged**